python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+) -> NoReturn:
+    # Do stuff
+```
+
+Notes:
+
+* Like all other SDK languages so far, we can't reasonably provide types for args on the "self" continue as new
+* More investigation is needed on whether I can stop the event loop inside the continue as new without it returning
+  * I probably can't so I probably need to make this call `async`, but I probably don't have to throw
+  * While I thought I could just stop the event loop and then return, that doesn't prevent the next line of code from
+    potentially setting local state that could change a query result
+  * This will be subject to implementation-time research. The goal is to try to never return from the function (if it's
+    `async` that implies never return from an `await` on it) and if possible do not raise an error.
+
+#### `temporalio.workflow.ContinueAsNewConfig`
+
+`TypedDict` of `continue_as_new` configuration options.
+
+#### `temporalio.workflow.get_external_workflow_handle`
+
+```python
+class ExternalWorkflowHandle(Generic[WorkflowType], Cancellable):
+
+    @property
+    def id(self) -> str:
+        pass
+
+    @property
+    def run_id(self) -> Optional[str]:
+        pass
+
+    # Include same typed overloads as WorkflowHandle
+    async def signal(self, signal: Any, *args: Any) -> None:
+        pass
+
+    async def cancel(self) -> None:
+        pass
+
+def get_external_workflow_handle_for(
+    workflow: Union[
+        Callable[[WorkflowType, Any], Awaitable[Any]],
+        Callable[[WorkflowType], Awaitable[Any]],
+    ],
+    workflow_id: str,
+    *,
+    run_id: Optional[str] = None,
+) -> ExternalWorkflowHandle[WorkflowType]:
+    return get_external_workflow_handle(workflow_id, run_id=run_id)
+
+def get_external_workflow_handle(
+    workflow_id: str,
+    *,
+    run_id: Optional[str] = None,
+) -> ExternalWorkflowHandle[Any, Any]:
+    # Do stuff
+```
+
+Notes:
+
+* This intentionally matches the client API
+
+#### `temporalio.workflow.upsert_search_attributes`
+
+```python
+def upsert_search_attributes(attrs: Mapping[str, Any]) -> None:
+    pass
+```
+
+Notes:
+
+* I prefer `set` over `upsert` here, but not enough to fight for it
+* Originally there was concern that I couldn't report data conversion failure on search attributes, but since custom
+  converters are not allowed, I can just make sync calls to the JSON data converter
+
+#### `temporalio.workflow` Async Utilities
+
+```python
+async def sleep(delay: float, result: T = None) -> T:
+  pass
+
+async def wait_condition(fn: Callable[[], Awaitable[bool]], *, timeout: Optional[float] = None):
+    pass
+```
+
+Notes:
+
+* See discussion in "Runtime - Workflow Stepped Execution" on whether to reuse `asyncio.sleep`
+* Like `temporal.activity.wait*` we choose the "wait_" prefix here as Python often does, e.g. for events which this is
+  most analogous to
+
+#### `temporalio.workflow` Miscellaneous
+
+```python
+
+@dataclass(frozen=True)
+class Info:
+    pass
+
+def is_replaying() -> bool:
```

Comment:
Let's put this under `unsafe` like TS and Java

Reply:
:+1:

Reply:
Done